crowbar_framework/spec/lib/crowbar/lock/shared_non_blocking_spec.rb

```diff
@@ -0,0 +1,177 @@
+require "spec_helper"
+require "timeout"
+require "rspec/expectations"
+
+RSpec::Matchers.define :be_unlocked do
+  match do |actual|
+    !actual.locked? && !File.exists?(actual.path)
+  end
+  failure_message do |actual|
+    actual.locked? ?
+      "expected that #{actual} would not be locked"
+    : "expected that #{actual.path} would not exist"
+  end
+end
+
+RSpec::Matchers.define :be_locked_by_me do
+  match do |actual|
+    actual.locked? && File.exists?(actual.path)
+  end
+  failure_message do |actual|
+    ! actual.locked? ?
+      "expected that #{actual} would be locked"
+    : "expected that #{actual.path} would exist"
+  end
+end
+
+RSpec::Matchers.define :be_locked_by_another do
+  match do |actual|
+    !actual.locked? && File.exists?(actual.path)
+  end
+  failure_message do |actual|
+    actual.locked? ?
+      "expected that #{actual} would not be locked"
+    : "expected that #{actual.path} would exist"
+  end
+end
+
+describe Crowbar::Lock::SharedNonBlocking do
+  def safe_run(cmd)
+    ret = system(cmd)
+    unless ret
+      raise "pid #{$?.pid} return returned #{$?.exitstatus} for command:\n#{cmd}"
+    end
+    ret
+  end
+
+  let(:klass) { Crowbar::Lock::SharedNonBlocking }
+  let(:local_executor) { lambda { |node, cmd| safe_run(cmd) } }
+  def racy_executor(start_delay, end_delay)
+    lambda do |node, cmd|
+      # Make the critical region inside the flock much longer.
+      # Note that the command to be run has already been logged at this point,
+      # so the modifications won't show up in log/test.log.
+      cmd.gsub!("-ec '\\\n", "\\0sleep #{start_delay}; ") if start_delay > 0
+      cmd.gsub!(/'$/, "; sleep #{end_delay}\\0") if end_delay > 0
+      safe_run(cmd)
+    end
+  end
+  let(:defaults) do
+    {
+      executor: local_executor,
+      node: 'dummy', # local_executor ignores this
+      owner: 'rspec',
+      reason: 'testing',
+    }
+  end
+  let(:lock) { new_lock }
+
+  def new_lock(options = {})
+    options[:path] ||= Dir::Tmpname.create(['test_shared_nb-', '.lock']) { }
```

Comment:
Style/SpaceInsideBlockBraces: Space inside empty braces detected.<br>Style/StringLiterals: Prefer double-quoted strings unless you need single quotes to avoid extra backslashes for escaping.